src/c/coap-server.c

```diff
@@ -0,0 +1,465 @@
+    goto finish;
+  }
+
+  /* generate and post an event with the data */
+  devsdk_commandresult results[1];
+  results[0].origin = 0;
+  results[0].value = iot_data;
+
+  devsdk_post_readings (sdk_ctx->service, device->name, resource->name, results);
+  iot_data_free (results[0].value);
+
+  response->code = COAP_RESPONSE_CODE (204);
+
+ finish:
+  if (device)
+  {
+    edgex_free_device (device);
+  }
+}
+
+int
+run_server (coap_driver *driver)
+{
+  coap_context_t  *ctx = NULL;
+  coap_address_t bind_addr;
+  coap_resource_t *resource = NULL;
+  int result = EXIT_FAILURE;
+  sdk_ctx = driver;
+  struct sigaction sa;
+
+  coap_startup ();
+
+  /* Use EdgeX log level */
+  coap_log_t log_level;
+  switch (sdk_ctx->lc->level)
+  {
+    case IOT_LOG_ERROR:
+      log_level = LOG_ERR;
+      break;
+    case IOT_LOG_WARN:
+      log_level = LOG_WARNING;
+      break;
+    case IOT_LOG_DEBUG:
+      log_level = LOG_DEBUG;
+      break;
+    default:
+      log_level = LOG_INFO;
+  }
+  coap_set_log_level (log_level);
+  /* workaround for tinydtls log level mismatch to avoid excessive debug logging */
+  coap_tls_version_t *ver = coap_get_tls_library_version ();
+  if (ver->type == COAP_TLS_LIBRARY_TINYDTLS && log_level == LOG_INFO)
+  {
+    coap_dtls_set_log_level (log_level - 1);
+  }
+  else
+  {
+    coap_dtls_set_log_level (log_level);
+  }
+
+  /* Resolve destination address where server should be sent. Use CoAP default ports. */
+  coap_proto_t proto = COAP_PROTO_UDP;
+  char *port = "5683";
+  if (driver->security_mode != SECURITY_MODE_NOSEC)
+  {
+    proto = COAP_PROTO_DTLS;
+    port = "5684";
+  }
+  if (resolve_address (iot_data_string (driver->coap_bind_addr), port, &bind_addr) < 0) {
+    iot_log_error (sdk_ctx->lc, "failed to resolve CoAP bind address");
+    goto finish;
+  }
+
+  /* setup libcoap for a server */
+  if (!(ctx = coap_new_context (NULL)))
+  {
+    iot_log_error (sdk_ctx->lc, "cannot initialize context");
+    goto finish;
+  }
+
+  if (driver->security_mode == SECURITY_MODE_PSK)
+  {
+    /* use iterator just to get address of PSK key data */
+    iot_data_array_iter_t array_iter;
+    iot_data_array_iter (driver->psk_key, &array_iter);
+    iot_data_array_iter_next(&array_iter);
+
+    if (!(coap_context_set_psk (ctx, "", (uint8_t *)iot_data_array_iter_value (&array_iter),
+                                iot_data_array_length (driver->psk_key))))
+    {
+      iot_log_error (sdk_ctx->lc, "cannot initialize PSK");
+      goto finish;
+    }
+  }
+
+  if (!(coap_new_endpoint (ctx, &bind_addr, proto)))
+  {
+    iot_log_error (sdk_ctx->lc, "cannot initialize listen endpoint");
+    goto finish;
+  }
+
+  /* Creates handler for PUT, which is not what we want... */
+  resource = coap_resource_unknown_init (&data_handler);
+  /* ... so add POST handler also. */
+  coap_register_handler (resource, COAP_REQUEST_POST, &data_handler);
```

Comment:
Hope the handlers for all the other methods like GET, PUT, DELETE will be added in the future. Please confirm.

Reply:
As a RESTful protocol I agree that CoAP naturally allows for extension of device and resource management out to remote sensors. Maybe we first need to know if users have that itch to scratch in the open source sense.